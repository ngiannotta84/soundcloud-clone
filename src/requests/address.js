// const address = "https://soundcloud-clone-api.onrender.com";
const address = "http://localhost:4000";

export default address;
