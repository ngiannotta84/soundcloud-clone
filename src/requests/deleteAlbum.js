import { deleteRequest } from "./helpers";

const deleteAlbum = async (id) => {
  const response = await deleteRequest("albums", id);
  return response;
};

export default deleteAlbum;
