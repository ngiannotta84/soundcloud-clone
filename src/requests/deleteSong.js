import { deleteRequest } from "./helpers";

const deleteSong = async (id) => {
  try {
    const response = await deleteRequest("songs", id);
    return response;
  } catch (err) {
    throw new Error(err);
  }
};

export default deleteSong;
