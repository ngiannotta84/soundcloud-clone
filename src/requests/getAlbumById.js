import { getByIdRequest } from "./helpers";

const getAlbumById = async (id) => {
  const response = await getByIdRequest("albums", id);
  return response;
};

export default getAlbumById;
