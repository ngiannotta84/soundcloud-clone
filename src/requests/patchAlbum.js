import { patchRequest } from "./helpers";

const patchAlbum = async (id, data) => {
  const response = await patchRequest("albums", id, data);
  return response;
};

export default patchAlbum;
