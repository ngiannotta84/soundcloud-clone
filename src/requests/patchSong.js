import { patchRequest } from "./helpers";

const patchSong = async (id, data) => {
  const response = await patchRequest("songs", id, data);
  return response;
};

export default patchSong;
