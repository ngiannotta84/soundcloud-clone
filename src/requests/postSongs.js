import { postRequest } from "./helpers";

const postSongs = async (data) => {
  if (
    !data.name ||
    !data.AlbumId ||
    data.position === undefined ||
    !data.audio
  ) {
    throw new Error("missing data");
  }

  const response = await postRequest("songs", data);
  return response;
};

export default postSongs;
