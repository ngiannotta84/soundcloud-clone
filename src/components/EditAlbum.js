import React, { useEffect, useState } from "react";
import { useParams, useNavigate } from "react-router-dom";
import { v4 as uuid } from "uuid";
import PropTypes from "prop-types";
import Alert from "./Alert";
import Loader from "./Loader";
import Confirm from "./Confirm";
import getAlbumById from "../requests/getAlbumById";
import deleteAlbum from "../requests/deleteAlbum";
import patchAlbum from "../requests/patchAlbum";
import patchSong from "../requests/patchSong";
import deleteSong from "../requests/deleteSong";
import postSongs from "../requests/postSongs";

const EditAlbum = ({ userName }) => {
  const initialState = {
    name: "",
    Songs: [],
  };

  const [placeHolders, setPlaceHolders] = useState(initialState);
  const [album, setAlbum] = useState({});
  const [editSongs, setEditSongs] = useState([]);
  const [newSongs, setNewSongs] = useState([]);
  const [alert, setAlert] = useState("");
  const [confirm, setConfirm] = useState("");
  const [loading, setLoading] = useState(false);

  const { albumId } = useParams();
  const navigate = useNavigate();

  const handleAlbumNameChange = (e) => {
    const { value } = e.target;
    setAlbum((prev) => {
      return { ...prev, name: value };
    });
  };

  const handleImageChange = (e) => {
    setAlert("");
    const file = e.target.files[0];
    if (file.type.split("/")[0] !== "image") {
      setAlert("Album art must be of type image");
    } else {
      setAlbum((prev) => {
        return { ...prev, image: file };
      });
    }
  };

  const handleSongNameChange = (e, i) => {
    const { name, value } = e.target;
    setEditSongs((prev) => {
      const clone = [...prev];
      clone[i][name] = value;
      return clone;
    });
  };

  const handleSongFileChange = (e, i) => {
    setAlert("");
    const file = e.target.files[0];
    if (file.type.split("/")[0] !== "audio") {
      setAlert("Songs files must be of type audio");
    } else {
      setEditSongs((prev) => {
        const clone = [...prev];
        clone[i].audio = file;
        return clone;
      });
    }
  };

  const handleSongDelete = (e, i) => {
    const { checked } = e.target;
    setEditSongs((prev) => {
      const clone = [...prev];
      clone[i].delete = checked;
      return clone;
    });
  };

  const handleNewSongNameChange = (e, i) => {
    const { name, value } = e.target;
    setNewSongs((prev) => {
      const clone = [...prev];
      clone[i][name] = value;
      return clone;
    });
  };

  const handleNewSongFileChange = (e, i) => {
    const file = e.target.files[0];
    setNewSongs((prev) => {
      const clone = [...prev];
      clone[i].audio = file;
      return clone;
    });
  };

  const addSong = () => {
    setNewSongs((prev) => [
      ...prev,
      {
        name: "",
        image: "",
        key: uuid(),
      },
    ]);
  };

  const handleDeleteSong = (i) => {
    setNewSongs((prev) => {
      const clone = [...prev];
      clone.splice(i, 1);
      return clone;
    });
  };

  const cancel = () => {
    navigate(-1);
  };

  const deleteAlbumClick = async () => {
    try {
      setLoading(true);
      await deleteAlbum(albumId);
      navigate(-1);
    } catch (err) {
      setAlert("failed to delete album");
    } finally {
      setLoading(false);
    }
  };

  const saveChanges = async () => {
    try {
      setAlert("");
      setLoading(true);

      const { length: newSongLength } = newSongs;
      for (let i = 0; i < newSongLength; i += 1) {
        const song = newSongs[i];
        if (!song.name || !song.audio) {
          setAlert("all new songs must contain a name and audio");
          return;
        }
      }

      const albumPromise = [];

      if (album.name || album.image) {
        const data = {
          name: album.name || undefined,
          image: album.image || undefined,
        };
        albumPromise.push(patchAlbum(albumId, data));
      }

      let position = 0;
      const songUpdatePromises = editSongs.map((song, i) => {
        if (song.delete) {
          return deleteSong(song.id);
        }
        const songPosition = Number(placeHolders.Songs[i].position);
        if (song.name || song.audio || position !== songPosition) {
          const data = {
            name: song.name || undefined,
            audio: song.audio || undefined,
            position: position === songPosition ? undefined : position,
          };
          position += 1;
          return patchSong(song.id, data);
        }
        position += 1;
        return null;
      });

      const songPostPromises = newSongs.map((song) => {
        const data = {
          name: song.name,
          audio: song.audio,
          position,
          AlbumId: albumId,
        };
        position += 1;
        return postSongs(data);
      });

      const promises = albumPromise.concat(
        songUpdatePromises,
        songPostPromises
      );

      await Promise.all(promises);

      navigate(`/profile/${userName}`);
    } catch (err) {
      setAlert(err.message);
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    (async () => {
      try {
        setLoading(true);
        const response = await getAlbumById(albumId);
        if (response) {
          setPlaceHolders(response);
          const emptySongArray = response.Songs.map((song) => {
            return {
              name: "",
              audio: "",
              delete: false,
              id: song.id,
              key: uuid(),
            };
          });
          setEditSongs(emptySongArray);
        } else {
          setPlaceHolders(initialState);
        }
      } catch (err) {
        setPlaceHolders(initialState);
        setAlert("failed to find Album");
      } finally {
        setLoading(false);
      }
    })();
  }, [albumId]);

  return (
    <div className="upload">
      <Loader loading={loading} />
      <Alert message={alert} />
      <form className="upload__form">
        <h2 className="upload__title--large">Edit Album</h2>
        <div className="upload__album">
          <label htmlFor="name" className="upload__label">
            <span className="upload-info">Album Name</span>
            <input
              type="text"
              id="name"
              value={album.name}
              placeholder={placeHolders.name}
              onChange={handleAlbumNameChange}
            />
          </label>
          <label htmlFor="image" className="upload__label">
            <span className="upload-info">Album Artwork</span>
            <input
              type="file"
              id="image"
              onChange={handleImageChange}
              className="upload__file"
            />
          </label>
        </div>
        <h3 className="upload__title--large">Edit Songs</h3>
        {editSongs.map((song, i) => {
          return (
            <div key={`${song.key}`} className="upload__song">
              <h3 className="upload__song__title">Song {i + 1}</h3>
              <label htmlFor={`song name ${i}`} className="upload__label">
                <span className="upload-info">Song Name</span>
                <input
                  type="text"
                  id={`song name ${i}`}
                  name="name"
                  value={editSongs[i].name}
                  placeholder={placeHolders.Songs[i].name}
                  onChange={(e) => handleSongNameChange(e, i)}
                  className="upload__text-input"
                />
              </label>
              <label htmlFor={`song image ${i}`} className="upload__label">
                <span className="upload-info">Audio</span>
                <input
                  type="file"
                  id={`song image ${i}`}
                  onChange={(e) => handleSongFileChange(e, i)}
                  className="upload__file"
                />
              </label>
              <label htmlFor={`delete song ${i}`}>
                <span className="upload-info">Delete Song?</span>
                <input
                  type="checkbox"
                  id={`delete song ${i}`}
                  name="delete"
                  checked={editSongs[i].delete}
                  onChange={(e) => handleSongDelete(e, i)}
                />
              </label>
            </div>
          );
        })}
        <h3 className="upload__title--large">Add Songs</h3>
        {newSongs.map((song, i) => {
          return (
            <div key={song.key} className="upload__song">
              <h3 className="upload__song__title">New Song {i + 1}</h3>
              <label htmlFor={`new song name ${i}`} className="upload__label">
                <span className="upload-info">Song Name</span>
                <input
                  type="text"
                  id={`new song name ${i}`}
                  name="name"
                  value={newSongs[i].name}
                  onChange={(e) => handleNewSongNameChange(e, i)}
                  className="upload__text-input"
                />
              </label>
              <label htmlFor={`new song image ${i}`} className="upload__label">
                <span className="upload-info">Audio</span>
                <input
                  type="file"
                  id={`new song image ${i}`}
                  onChange={(e) => handleNewSongFileChange(e, i)}
                  className="upload__file"
                />
              </label>
              <button type="button" onClick={() => handleDeleteSong(i)}>
                Remove Song
              </button>
            </div>
          );
        })}
        <button type="button" onClick={addSong} className="upload__button">
          Add Song
        </button>
        <div>
          <button
            type="button"
            onClick={() => setConfirm("cancel")}
            className="upload__button"
          >
            Cancel
          </button>
          <button
            type="button"
            onClick={() => setConfirm("delete")}
            className="upload__button"
          >
            Delete Album
          </button>
          <button
            type="button"
            onClick={() => setConfirm("save")}
            className="upload__button"
          >
            Save Changes
          </button>
        </div>
      </form>
      {confirm === "cancel" && (
        <Confirm callback={cancel} setConfirm={setConfirm} />
      )}
      {confirm === "delete" && (
        <Confirm callback={deleteAlbumClick} setConfirm={setConfirm} />
      )}
      {confirm === "save" && (
        <Confirm callback={saveChanges} setConfirm={setConfirm} />
      )}
    </div>
  );
};

EditAlbum.defaultProps = {
  userName: "",
};

EditAlbum.propTypes = {
  userName: PropTypes.string,
};

export default EditAlbum;
