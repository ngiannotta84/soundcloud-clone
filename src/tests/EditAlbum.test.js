import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { act } from "react-dom/test-utils";
import EditAlbum from "../components/EditAlbum";
import * as getAlbumById from "../requests/getAlbumById";
import * as deleteAlbum from "../requests/deleteAlbum";
import * as patchAlbum from "../requests/patchAlbum";
import * as patchSong from "../requests/patchSong";
import * as deleteSong from "../requests/deleteSong";
import * as postSongs from "../requests/postSongs";
import fakeAlbumData from "./testData/fakeAlbumData";

describe("Edit", () => {
  const deleteAlbumMock = jest.fn();
  const deleteSongMock = jest.fn();
  const patchAlbumMock = jest.fn();
  const patchSongMock = jest.fn();
  const postSongsMock = jest.fn();
  jest.spyOn(getAlbumById, "default").mockResolvedValue(fakeAlbumData[0]);
  jest.spyOn(deleteAlbum, "default").mockImplementation(deleteAlbumMock);
  jest.spyOn(deleteSong, "default").mockImplementation(deleteSongMock);
  jest.spyOn(patchAlbum, "default").mockImplementation(patchAlbumMock);
  jest.spyOn(patchSong, "default").mockImplementation(patchSongMock);
  jest.spyOn(postSongs, "default").mockImplementation(postSongsMock);

  test("snapshot", async () => {
    let asFragment;
    await act(() => {
      const view = render(
        <Router>
          <EditAlbum />
        </Router>
      );
      asFragment = view.asFragment;
    });

    expect(asFragment()).toMatchSnapshot();
  });

  describe("tests", () => {
    beforeEach(async () => {
      await act(() => {
        render(
          <Router>
            <EditAlbum />
          </Router>
        );
      });
    });

    test("renders correctly", async () => {
      expect(screen.getByText(/^edit$/i)).toBeInstanceOf(HTMLHeadingElement);
      expect(screen.getByText(/edit album info/i)).toBeInstanceOf(
        HTMLHeadingElement
      );
      expect(screen.getByLabelText(/album name/i)).toBeInTheDocument();
      expect(screen.getByLabelText(/album artwork/i)).toBeInTheDocument();
      expect(screen.getByText(/edit songs/i)).toBeInstanceOf(
        HTMLHeadingElement
      );
      const addSongs = screen.getAllByText(/add song/i);
      expect(addSongs[0]).toBeInstanceOf(HTMLHeadingElement);
      expect(addSongs[1]).toBeInstanceOf(HTMLButtonElement);
      expect(screen.getByText(/cancel/i)).toBeInstanceOf(HTMLButtonElement);
      expect(screen.getByText(/delete album/i)).toBeInstanceOf(
        HTMLButtonElement
      );
      expect(screen.getByText(/save changes/i)).toBeInstanceOf(
        HTMLButtonElement
      );
    });

    test("add song", () => {
      const addSongButton = screen.getAllByText(/add song/i)[1];
      fireEvent.click(addSongButton);
      const inputs = screen.getAllByLabelText(/new song/i);

      expect(inputs[0]).toHaveAttribute("type", "text");
      expect(inputs[1]).toHaveAttribute("type", "file");
    });
  });
});
