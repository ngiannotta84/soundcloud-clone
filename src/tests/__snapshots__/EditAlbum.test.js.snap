// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit snapshot 1`] = `
<DocumentFragment>
  <div>
    <span
      aria-label="Loading Spinner"
      style="width: 150px; height: 150px; border-radius: 100%; border: 2px solid; border-top-color: #B133B1; border-bottom-color: transparent; border-left-color: #B133B1; border-right-color: #B133B1; display: inline-block; animation: react-spinners-ClipLoader-clip 0.75s 0s infinite linear; animation-fill-mode: both;"
    />
    <form>
      <h2>
        Edit
      </h2>
      <h3>
        Edit Album Info
      </h3>
      <label
        for="name"
      >
        <span>
          Album Name
        </span>
        <input
          id="name"
          placeholder=""
          type="text"
          value=""
        />
      </label>
      <label
        for="image"
      >
        <span>
          Album Artwork
        </span>
        <input
          id="image"
          type="file"
        />
      </label>
      <h3>
        Edit Songs
      </h3>
      <h3>
        Add Songs
      </h3>
      <button
        type="button"
      >
        Add Song
      </button>
      <div>
        <button
          type="button"
        >
          Cancel
        </button>
        <button
          type="button"
        >
          Delete Album
        </button>
        <button
          type="button"
        >
          Save Changes
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
